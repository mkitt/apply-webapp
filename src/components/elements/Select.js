// @flow
import React, { type Node } from 'react'
import glamorous from 'glamorous'
import { COLORS } from '../../constants/theme'
import { SelectChevronIcon } from './Icons'

type Props = {
  children: Node,
  label: string,
  name: string,
  onChange: Function,
  value: string,
}

export default ({ children, label, name, onChange, value, ...props }: Props) => (
  <Label htmlFor={name}>
    <Select
      aria-label={label}
      id={name}
      name={name}
      onChange={onChange}
      value={value}
      {...props}
    >
      { children }
    </Select>
    <SelectIcon>
      <SelectChevronIcon />
    </SelectIcon>
  </Label>
)

// -------------------------------------

const Label = glamorous.label({
  position: 'relative',
})

const Select = glamorous.select({
  WebkitAppearance: 'button',
  WebkitBorderRadius: 0,
  WebkitBoxShadow: 'none',
  WebkitPaddingEnd: 0,
  WebkitPaddingStart: 0,
  WebkitUserSelect: 'none',
  width: '100%',
  height: 48,
  margin: 0,
  padding: '0 32px',
  overflow: 'hidden',
  fontSize: 14,
  fontFamily: 'inherit',
  textTransform: 'uppercase',
  color: COLORS.WHITE,
  backgroundColor: COLORS.GREY_4,
  backgroundImage: 'none',
  backgroundPosition: 'initial',
  backgroundRepeat: 'no-repeat',
  borderRadius: 0,
  border: 0,
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap',
  ':focus': {
    borderColor: COLORS.GREY_9,
    outline: 0,
  },
})

const SelectIcon = glamorous.div({
  position: 'absolute',
  top: 'calc(50% - 8px)',
  right: 32,
  color: COLORS.WHITE,
  pointerEvents: 'none',
})
