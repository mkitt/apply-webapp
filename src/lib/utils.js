// @flow
import { __, addIndex, always, and, complement, either, gt, isEmpty, isNil, map } from 'ramda'
import { withProps } from 'recompose'

// -------------------------------------
// Environments
export const isDev = always(process.env.NODE_ENV === 'development')

// -------------------------------------
// Types
export const isNotNil = complement(isNil)
export const isNotEmpty = complement(isEmpty)
export const isEmptyOrNil = either(isNil, isEmpty)
export const isNotEmptyAndNotNil = and(isNotNil, isNotEmpty)

// -------------------------------------
// Relation
export const gt0 = gt(__, 0)

// -------------------------------------
// Lists
export const nool = always([])()
export const mapIndexed = addIndex(map)

// -------------------------------------
// Objects
export const noob = always({})()

// -------------------------------------
// Functions
export const noop = () => {}

export const log = withProps(props => (
  // eslint-disable-next-line no-console
  console.log(props.displayName || 'üëç', props) && props
))

export const warn = withProps(props => (
  // eslint-disable-next-line no-console
  console.warn(props.displayName || '‚ö†Ô∏è', props) && props
))
