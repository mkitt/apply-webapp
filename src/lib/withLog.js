import { withProps } from 'recompose'

export default withProps(props => (
  // eslint-disable-next-line
  console.log(props.logName || props.displayName || '👍', props) && props
))
