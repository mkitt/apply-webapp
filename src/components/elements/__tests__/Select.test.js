import React from 'react'
import renderer from 'react-test-renderer'
import Select from '../Select'
import { noop } from '../../../lib/utils'

describe('<Select />', () => {
  it('shows the correct snapshot tree for the <Select> element', () => {
    const component = (
      <Select label="Score" name="select-name" onChange={noop} value={null}>
        <option value={null}>Score</option>
        <option value={0}>0 - Bad</option>
        <option value={1}>1 - Meh</option>
        <option value={2}>2 - Good</option>
        <option value={3}>3 - Excellent</option>
      </Select>
    )
    const tree = renderer.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
