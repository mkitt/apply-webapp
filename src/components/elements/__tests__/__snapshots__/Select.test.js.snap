// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> shows the correct snapshot tree for the <Select> element 1`] = `
.css-0,
[data-css-0] {
  position: relative;
}

.css-1,
[data-css-1] {
  -webkit-appearance: button;
  -webkit-border-radius: 0;
  -webkit-box-shadow: none;
  -webkit-padding-end: 0;
  -webkit-padding-start: 0;
  -webkit-user-select: none;
  width: 100%;
  height: 48px;
  margin: 0;
  padding: 0 32px;
  overflow: hidden;
  font-size: 14px;
  font-family: inherit;
  text-transform: uppercase;
  color: #fff;
  background-color: #4d4d4d;
  background-image: none;
  background-position: initial;
  background-repeat: no-repeat;
  border-radius: 0;
  border: 0;
  text-overflow: ellipsis;
  white-space: nowrap;
}

.css-1:focus,
[data-css-1]:focus {
  border-color: #999;
  outline: 0;
}

.css-3,
[data-css-3] {
  position: absolute;
  top: calc(50% - 8px);
  right: 32px;
  color: #fff;
  pointer-events: none;
}

<label
  className="css-0"
  htmlFor="select-name"
>
  <select
    aria-label="Score"
    className="css-1"
    id="select-name"
    name="select-name"
    onChange={[Function]}
    value={null}
  >
    <option
      value={null}
    >
      Score
    </option>
    <option
      value={0}
    >
      0 - Bad
    </option>
    <option
      value={1}
    >
      1 - Meh
    </option>
    <option
      value={2}
    >
      2 - Good
    </option>
    <option
      value={3}
    >
      3 - Excellent
    </option>
  </select>
  <div
    className="css-3"
  >
    <svg
      className="css-nil"
      height={12}
      viewBox="0 0 21 12"
      width={21}
    >
      <g
        fill="none"
        fillRule="evenodd"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
      >
        <polyline
          points="19.182 1.067 10.611 10.067 1.182 1.067"
        />
      </g>
    </svg>
  </div>
</label>
`;
